sdk/trace/sampling.go

```diff
@@ -64,56 +68,258 @@ type SamplingResult struct {
+		state = mod
+	} else {
+		otel.Handle(fmt.Errorf("could not update tracestate: %q", err))
+	}
 	return SamplingResult{
-		Decision:   Drop,
-		Tracestate: psc.TraceState(),
+		Decision:   RecordAndSample,
+		Tracestate: state,
 	}
 }
 
+// combineTracestate combines an existing OTel tracestate fragment,
+// which is the value of a top-level "ot" tracestate vendor tag.
+func combineTracestate(incoming, updated string) string {
+	// `incoming` is formatted according to the OTel tracestate
+	// spec, with colon separating two-byte key and value, with
+	// semi-colon separating key-value pairs.
+	//
+	// `updated` should be a single two-byte key:value to modify
+	// or insert therefore colonOffset is 2 bytes, valueOffset is
+	// 3 bytes into `incoming`.
+	const colonOffset = 2
+	const valueOffset = colonOffset + 1
+
+	if incoming == "" {
+		return updated
+	}
+	var out strings.Builder
+
+	// The update is expected to be a single key-value of the form
+	// `XX:value` for with two-character key.
+	upkey := updated[:colonOffset]
+
+	// In this case, there is an existing field under "ot" and we
+	// need to combine.  We will pass the parts of "incoming"
+	// through except the field we are updating, which we will
+	// modify if it is found.
+	foundUp := false
+
+	for count := 0; len(incoming) != 0; count++ {
+		key, rest, hasCol := strings.Cut(incoming, ":")
+		if !hasCol {
+			// return the updated value, ignore invalid inputs
+			return updated
+		}
+		value, next, _ := strings.Cut(rest, ";")
+
+		if key == upkey {
+			value = updated[valueOffset:]
+			foundUp = true
+		}
+		if count != 0 {
+			out.WriteString(";")
+		}
+		out.WriteString(key)
+		out.WriteString(":")
+		out.WriteString(value)
+
+		incoming = next
+	}
+	if !foundUp {
+		out.WriteString(";")
+		out.WriteString(updated)
+	}
+	return out.String()
+}
+
 func (ts traceIDRatioSampler) Description() string {
 	return ts.description
 }
 
+const (
+	// DefaultSamplingPrecision is the number of hexadecimal
+	// digits of precision used to expressed the samplling probability.
+	DefaultSamplingPrecision = 4
+
+	// MinSupportedProbability is the smallest probability that
+	// can be encoded by this implementation, and it defines the
+	// smallest interval between probabilities across the range.
+	// The largest supported probability is (1-MinSupportedProbability).
+	//
+	// This value corresponds with the size of a float64
+	// significand, because it simplifies this implementation to
+	// restrict the probability to use 52 bits (vs 56 bits).
+	minSupportedProbability float64 = 1 / float64(maxAdjustedCount)
+
+	// maxSupportedProbability is the number closest to 1.0 (i.e.,
+	// near 99.999999%) that is not equal to 1.0 in terms of the
+	// float64 representation, having 52 bits of significand.
+	// Other ways to express this number:
+	//
+	//   0x1.ffffffffffffe0p-01
+	//   0x0.fffffffffffff0p+00
+	//   math.Nextafter(1.0, 0.0)
+	maxSupportedProbability float64 = 1 - 0x1p-52
```

Comment:
The name `maxSupportedProbability` could be misleading here, as consistent sampling naturally also supports sampling with a probability of 1.